tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,115 +1,373 @@
+                wavelength=wavelength,
+                pts_per_wavelength=self.pts_per_wavelength
+                )
+            self.origin = [sum(x) for x in zip(self.origin, surface.mon.center)]
+
+        self.origin[:] = [x / len(self.surfaces) for x in self.origin]
+        self.initialized = True
+
+    # pylint:disable=too-many-locals
+    def _radiation_vectors_for_surface(self, theta: float, phi: float, data: Near2FarSurface):
+        """Compute radiation vectors at an angle in spherical coordinates 
+        for a given :class:`Near2FarSurface` and observation angles
 
         Parameters
         ----------
         theta : float
-            Polar angle (rad) downward from x=y=0 line.
+            Polar angle (rad) downward from x=y=0 line relative to monitor center.
         phi : float
-            Azimuthal (rad) angle from x=z=0 line.
+            Azimuthal (rad) angle from y=z=0 line relative to monitor center.
+        data : :class:`Near2FarSurface`
+            Data set corresponding to a single surface monitor.
 
         Returns
         -------
         tuple[float, float, float, float]
-            ``N_theta``, ``N_phi``, ``L_theta``, ``L_phi`` radiation vectors.
+            ``N_theta``, ``N_phi``, ``L_theta``, ``L_phi`` radiation vectors for the given surface.
         """
 
-        # precompute trig functions and add extra dimensions
+        # precompute trig functions
         sin_theta = np.sin(theta)
         cos_theta = np.cos(theta)
         sin_phi = np.sin(phi)
         cos_phi = np.cos(phi)
 
-        # precompute fourier transform phase term {dx dy e^(ikrcos(psi))}
-        phase_x = np.exp(1j * self.k0 * self.xx * sin_theta * cos_phi)
-        phase_y = np.exp(1j * self.k0 * self.yy * sin_theta * sin_phi)
-        phase = self.dx * self.dy * phase_x * phase_y
-
-        Jx_k = np.sum(self.Jx * phase)
-        Jy_k = np.sum(self.Jy * phase)
-        Mx_k = np.sum(self.Mx * phase)
-        My_k = np.sum(self.My * phase)
+        k_m = self.k_m
+
+        if data.axis == 0:
+            source_indices = [1,2]
+            source_names = ['y', 'z']
+        elif data.axis == 1:
+            source_indices = [0,2]
+            source_names = ['x', 'z']
+        else:
+            source_indices = [0,1]
+            source_names = ['x', 'y']
+
+        # make sure that observation points are interpreted w.r.t. the monitors' centroid
+        pts = [None] * 3
+        pts[0] = data.currents['x'] - self.origin[0]
+        pts[1] = data.currents['y'] - self.origin[1]
+        pts[2] = data.currents['z'] - self.origin[2]
+
+        phase_x = np.exp(-self.phasor_sign * 1j * k_m * pts[0] * sin_theta * cos_phi)
+        phase_y = np.exp(-self.phasor_sign * 1j * k_m * pts[1] * sin_theta * sin_phi)
+        phase_z = np.exp(-self.phasor_sign * 1j * k_m * pts[2] * cos_theta)
+        phase = phase_x * phase_y * phase_z
+
+        J = [0, 0, 0]
+        M = [0, 0, 0]
+
+        J[source_indices[0]] = np.trapz(np.trapz(
```

Comment:
Any way to make this more readable?  It's tough to unpack.  
For example, defining a small function in this scope that integrates a single component (0, 1) of a current source data?  Any ideas?